pkg/controller/kubelet-config/helpers.go

```diff
@@ -110,13 +133,87 @@ func updateOriginalKubeConfigwithNodeConfig(node *osev1.Node, originalKubeletCon
 		originalKubeletConfig.NodeStatusUpdateFrequency = metav1.Duration{Duration: osev1.LowNodeStatusUpdateFrequency}
 	case osev1.DefaultUpdateDefaultReaction:
 		originalKubeletConfig.NodeStatusUpdateFrequency = metav1.Duration{Duration: osev1.DefaultNodeStatusUpdateFrequency}
+	case emptyInput:
+		return nil
 	default:
 		return fmt.Errorf("unknown worker latency profile type found %v, failed to update the original kubelet configuration", node.Spec.WorkerLatencyProfile)
 	}
 	// The kubelet configuration can be updated based on the cgroupmode as well here.
 	return nil
 }
 
+// isTechPreviewNoUpgradeEnabled returns a boolean accordingly if a feature is enabled
+// (TODO) This can be generically used in future by passing any feature as an argument.
+func isTechPreviewNoUpgradeEnabled(features *osev1.FeatureGate) bool {
+	if features == nil {
+		return false
+	}
+	if features.Spec.FeatureGateSelection.FeatureSet == osev1.TechPreviewNoUpgrade {
+		featureSet, ok := osev1.FeatureSets[osev1.TechPreviewNoUpgrade]
+		if !ok {
+			return false
+		}
+		for _, val := range featureSet.Enabled {
+			if val == cgroupv2Feature {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// updateMachineConfigwithCgroup updates the Machine Config object based on the cgroup mode present in the Config Node resource.
+func updateMachineConfigwithCgroup(node *osev1.Node, mc *mcfgv1.MachineConfig) error {
+	if node == nil {
+		return fmt.Errorf("node configuration not found, failed to update the machine config with the cgroup information")
+	}
+	if reflect.DeepEqual(node.Spec, osev1.NodeSpec{}) {
+		return fmt.Errorf("empty config node resource spec found")
+	}
+	if mc == nil || reflect.DeepEqual(mc.Spec, mcfgv1.MachineConfigSpec{}) {
+		return fmt.Errorf("machine config not found, failed to update the machine config with the cgroup information")
+	}
+	// updating the Machine Config resource with the relevant cgroup config
+	var (
+		kernelArgsv1       = []string{"systemd.unified_cgroup_hierarchy=0", "systemd.legacy_systemd_cgroup_controller=1"}
+		kernelArgsv2       = []string{"systemd.unified_cgroup_hierarchy=1", "cgroup_no_v1=\"all\"", "psi=1"}
+		kernelArgstoRemove []string
+		kernelArgstoAdd    []string
+	)
+	switch node.Spec.CgroupMode {
+	case osev1.CgroupModeV1:
+		kernelArgstoAdd = append(kernelArgstoAdd, kernelArgsv1...)
+		kernelArgstoRemove = append(kernelArgstoRemove, kernelArgsv2...)
+	case osev1.CgroupModeV2:
+		kernelArgstoAdd = append(kernelArgstoAdd, kernelArgsv2...)
+		kernelArgstoRemove = append(kernelArgstoRemove, kernelArgsv1...)
+	case emptyInput:
+		return nil
+	default:
+		return fmt.Errorf("unknown cgroup mode found %v, failed to update the machine config resource", node.Spec.CgroupMode)
+	}
+	for _, arg := range kernelArgstoRemove {
+		for key, value := range mc.Spec.KernelArguments {
+			if arg == value {
+				mc.Spec.KernelArguments = append(mc.Spec.KernelArguments[:key], mc.Spec.KernelArguments[key+1:]...)
+			}
+		}
```

Comment:
Hmm, we might end up out of bounds on this slice on removal if  `KernelArguments` isn't the length we think it will be. 

Thoughts on something like the following where we "build" the desired list instead of trying to "splice" the existing one? 

```
	// start with nothing
	var adjustedKernelArgs []string

	for _, arg := range mc.Spec.KernelArguments {
		// only append the args we want to keep, omitting the undesired
		if !ctrlcommon.InSlice(arg, kernelArgsToRemove) {
			adjustedKernelArgs = append(adjustedKernelArgs, arg)
		}
	}

	for _, arg := range kernelArgsToAdd {
		// add the additional that aren't already there
		if !ctrlcommon.InSlice(arg, adjustedKernelArgs) {
			adjustedKernelArgs = append(adjustedKernelArgs, arg)
		}
	}
       
        // overwrite at the end 
	mc.Spec.KernelArguments = adjustedKernelArgs
```
nit: I also feel like I want to capitalize the "to" in kernelArgs**to**Add/Remove, but I'll probably live :smile:

Reply:
Hey, Thanks for the suggestion.
I agree that the splicing may be slower compared to overwriting with the already built adjustedArgs, which in turn may cause the out of bound problem.
Addressed.